### DISCUSSION TITLE
TextInput AutoFocus issues

### DISCUSSION THREAD

**Discussion by saadlatif75** [react-native-tvos/react-native-tvos#753] [General Discussions]
### Description

How i can Focus Textinput in react native tv os i'm not be able to focus the textinput in react native

### React Native Version

0.73.6-1

### Output of `npx react-native info`

npx react-native info 
info Fetching system and libraries information...
System:
  OS: Windows 11 10.0.22631
  CPU: "(6) x64 AMD Ryzen 5 4500U with Radeon Graphics         "
  Memory: 1.75 GB / 15.40 GB
Binaries:
  Node:
    version: 18.19.0
    path: C:\Program Files\nodejs\node.EXE
  Yarn:
    version: 1.22.19
    path: ~\AppData\Roaming\npm\yarn.CMD
  npm:
    version: 8.19.3
    path: C:\Program Files\nodejs\npm.CMD
  Watchman: Not Found
SDKs:
  Android SDK:
    Android NDK: 22.1.7171670
  Windows SDK:
    AllowDevelopmentWithoutDevLicense: Enabled
    Versions:
      - 10.0.22621.0
IDEs:
  Android Studio: Not Found
  Visual Studio:
    - 17.8.34408.163 (Visual Studio Community 2022)
Languages:
  Java: 17.0.10
  Ruby: Not Found
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native: Not Found
  react-native-tvos:
    installed: 0.73.6-1
  react-native-windows: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: Not found
  newArchEnabled: Not found

info React Native v0.73.6 is now available (your project is running on v0.73.6-1).
info Changelog: https://github.com/facebook/react-native/releases/tag/v0.73.6
info Diff: https://react-native-community.github.io/upgrade-helper/?from=0.73.6        
info For more info, check out "https://reactnative.dev/docs/upgrading?os=windows".     
PS D:\tvhubRepo\tvhub> 

### Steps to reproduce

Use Textinput component in  the component not using remote control auto fous not work

**Comment by douglowder** [comment]
Please provide example code that reproduces this issue.

**Comment by saadlatif75** [comment]
> Please provide example code that reproduces this issue.
I want to focus textinput using tv remote control

```jsx
function CustomTextInput({
  leftIcon,
  label,
  inputStyle,
  isRightVisible,
  focusRef,
  isLabelVisible = true,
  textinputWidth = widthPixel(200),
  inputContainerStyle,
  marginHorizontal,
  marginVertical = heightPixel(30),
  onFocus,
  onBlur,
  ...props
}: any) {
  const [isPasswordVisible, setPasswordVisible] = useState(false);
  const [isFocused, setIsFocused] = useState(false); // Track focus state

  const eyeIcon = isPasswordVisible
    ? commonIcons.openEyeIcon
    : commonIcons.closeEyeIcon;

  const togglePasswordVisibility = () => {
    setPasswordVisible(!isPasswordVisible);
  };

  const textRef = useRef<TextInput>(null);
  useEffect(() => {
    if (focusRef) {
      focusRef.current = () => {
        textRef.current?.focus();
      };
    }
  }, [focusRef]);

  return (
    <View
      style={[
        styles.container,
        {
          width: textinputWidth,
          marginHorizontal: marginHorizontal,
          marginVertical: marginVertical,
        },
      ]}>
      {isLabelVisible ? (
        <View style={styles.labelContainer}>
          <Text style={styles.label}>{label}</Text>
        </View>
      ) : null}

      <TouchableOpacity
        activeOpacity={1}
        onFocus={() => {
          setIsFocused(true);
          onFocus && onFocus();
        }}
        onBlur={() => {
          setIsFocused(false);
          onBlur && onBlur();
        }}
        style={[
          styles.textInputContainer,
          inputContainerStyle,
          isFocused && styles.focusedInputContainer, // Apply focused styles
        ]}>
        <Image resizeMode="contain" style={styles.leftIcon}
source={leftIcon} />
        <TextInput
          style={[styles.input, inputStyle]}
          secureTextEntry={!isPasswordVisible && isRightVisible}
          {...props}
        />
        {isRightVisible && (
          <TouchableOpacity onPress={togglePasswordVisibility}>
            <Image
              resizeMode="contain"
              style={styles.leftIcon}
              source={eyeIcon}
            />
          </TouchableOpacity>
        )}
      </TouchableOpacity>
    </View>
  );
}
```

Second code
```jsx
<KeyboardAvoidingView style={commonContainerStyle} behavior="padding">
      <TVFocusGuideView autoFocus>
        <ScrollView keyboardDismissMode="interactive">
          <TopHeader centerText={'Choice Hub'} />
          <Text style={styles.registrationText}>Login</Text>
          <Text style={styles.enterYourDetailsText}>
            Enter your details to get Access of your account
          </Text>
          <CustomTextInput
            leftIcon={commonIcons.emailIcon}
            placeholder={'Enter your email address'}
            placeholderTextColor={appColor.grey200}
            label={'Email Address'}
            onSubmitEditing={() => {
              // Call the onFocusNextInput prop if it's provided
              alert('submit');
            }}
            onFocus={() => {
              // Your onFocus logic here
              console.log('Email input focused');
            }}
            onBlur={() => {
              // Your onBlur logic here
              console.log('Email input blurred');
            }}
            onChangeText={txt => {}}
          />
          <CustomTextInput
            leftIcon={commonIcons.lockIcon}
            placeholder={'Password'}
            placeholderTextColor={appColor.grey200}
            label={'Enter your password'}
            isRightVisble={true}
            onSubmitEditing={() => {
              // Call the onFocusNextInput prop if it's provided
              alert('submit');
            }}
            onFocus={() => {
              // Your onFocus logic here
              console.log('PAsss input focused');
            }}
            onBlur={() => {
              // Your onBlur logic here
              console.log('PAss input blurred');
            }}
            onChangeText={txt => {}}
          />
          <TouchableOpacity
            onPress={() => {
              navigation.navigate('ForgotPassword');
            }}
            style={styles.forgotPasswordButton}>
            <Text style={styles.forgotPasswordText}>Forgot password ?</Text>
          </TouchableOpacity>
          <ReuseableButton
            backgroundColor={appColor.skyblue100}
            buttonCenterText={'Login'}
            onButtonPress={() => {
              alert('d');
            }}
          />
          <TouchableOpacity
            style={{marginBottom: heightPixel(20)}}
            onPress={() => {
              navigation.navigate('Login');
            }}>
            <Text style={styles.alreadyhaveAccountText}>
              {commonText.dontHaveAccount}
              <Text style={styles.loginText}>Register</Text>
            </Text>
          </TouchableOpacity>
        </ScrollView>
      </TVFocusGuideView>
    </KeyboardAvoidingView>
```

**Comment by douglowder** [comment]
@saadlatif75 thanks for the example code. Are you having this problem on Apple TV, or Android TV? Also, your example code has a lot of external dependencies (styles, icons, other components), and appears to include navigation, so it would be helpful to get a complete working GitHub repo.

**Comment by douglowder** [comment]
@saadlatif75 in the meantime, this is how I wrap a TextInput with a TouchableOpacity and direct focus to it: https://github.com/react-native-tvos/TVInputDemo/blob/ea6abdba79c14c7554025efe769d72468019bd8c/src/common/StyledComponents.tsx#L129

**Comment by saadlatif75** [comment]
I see what the issue here when i wrap my whole component in the
TVEventhandler and pass auto focus props the after selecting the textinput
and typing some text in the textinput then I'm not be able to focus back
can you help me in this how i can auto focus without TVEventhandler React
native api

**Comment by douglowder** [comment]
@saadlatif75 I would need to see a complete working example of what you are trying to do.

**Comment by saadlatif75** [comment]
Ok, I will try to create a login screen, registration screen and OTP code
screen.Now these screen contain Textinput and button now for focus
textinput i wrap this textinput in Touchableopacity and using useref hook i
will open the Textinput then keyboard is open. Now after entering alphabet
when i submit the text then keyboard got close. Now issue is that when the
keyboard got close I'm not be able to focused other Textinput or button i
need to reload the whole component then it's working fine. I wrap my whole
login component in <Tvfocusguideview autofocus>

**Comment by donchop** [comment]
> @saadlatif75 in the meantime, this is how I wrap a TextInput with a TouchableOpacity and direct focus to it: https://github.com/react-native-tvos/TVInputDemo/blob/ea6abdba79c14c7554025efe769d72468019bd8c/src/common/StyledComponents.tsx#L129

@douglowder In your example,  when you use <WrappedTextInput/>  at android OS, you typing some text and click enter and after this your focus will gone to the first element in the screen? Is there a way to leave focus on the wrapper of current input?

**Comment by front1wss** [comment]
hi @saadlatif75 , did you manage this issue yet, seems like i have same problem?, i have login screen with 2 inputs and 1 button,
the issue appear when I wrapp it via ScrollView component (cant focus on a input or button using controller), if change ScrollView to View, it will work good

**Comment by saadlatif75** [comment]
Yes i write a custom logic to resolve this issue @front1wss 
f

**Comment by front1wss** [comment]
@saadlatif75 can you share with custom logic?

**Comment by saadlatif75** [comment]
@front1wss i will  use tvEventHandler api and by using UseRef hook i will handled this issue let me know when you avaible i will show you a live demo

**Comment by bilalhameed1** [comment]
> > @saadlatif75 in the meantime, this is how I wrap a TextInput with a TouchableOpacity and direct focus to it: https://github.com/react-native-tvos/TVInputDemo/blob/ea6abdba79c14c7554025efe769d72468019bd8c/src/common/StyledComponents.tsx#L129
> 
> @douglowder In your example, when you use at android OS, you typing some text and click enter and after this your focus will gone to the first element in the screen? Is there a way to leave focus on the wrapper of current input?

@douglowder I'm facing this same issue.
@donchop have you found any fix for this?

**Comment by donchop** [comment]
@bilalhameed1  Hello, nope.
I did logic that on every onEndEditing, onBlur props on TextInput i change states, it's quite ugly and have blink ( cose focus changed to first elemnt and than to current).

For example, if we have 2 text inputs for email and password:
```

your page:
  const [focusEmail, setFocusEmail] = useState<boolean>(false)
  const [focusPassword, setFocusPassword] = useState<boolean>(false)
  
       <YourTextInput
        onEndEditing={() => {
          setFocusEmail(true)
        }}
        onBlur={() => setFocusEmail(false)}
        focusWrapper={focusEmail}
        placeholderTx="common.email"
      />
      


YourTextInput:
interface IYourTextInputProps extends TextInputProps {
  focusWrapper?: boolean
}

const YourTextInput: FC<IYourTextInputProps> = ({
  focusWrapper = false,
  ...rest
}): JSX.Element => {
  const inputRef = useRef<TextInput>(null)
  const [isFocused, setIsFocused] = useState(false)

  const handleFocus = () => {
    setIsFocused(true)
  }

  const handleBlur = () => {
    setIsFocused(false)
  }

  return (
    <TouchableOpacity
      style={[S.WRAPPER, isFocused && S.CTR_FOCUS]}
      activeOpacity={1}
      onPress={() => inputRef?.current?.focus()}
      onBlur={handleBlur}
      hasTVPreferredFocus={focusWrapper}
    >
      <View style={[S.CONTAINER, styleContainer, !!error && S.CTR_ERROR]}>
        <View style={S.INPUT_CTR}>
          <TextInput
            autoCapitalize="none"
            onFocus={handleFocus}
            {...rest}
            cursorColor={colors.primary}
            selectionColor={colors.primary}
            ref={inputRef}
          />
        </View>
      </View>
    </TouchableOpacity>
  )
}

  
```

like this

**Comment by saadlatif75** [comment]
@donchop @douglowder @front1wss @bilalhameed1 Can you help me when i load data from an api then the focus got slow how i can resolve this i'm using onfoucs,onBlur method

**Comment by donchop** [comment]
@saadlatif75 It's hard to help without seeing your code, can you give an example of code where we can see this problem? A request for data, for example, you can hardcoded it in this example

**Comment by douglowder** [comment]
Moving this to the general discussions section of the repo.
